### PR TITLE
feat(feedback): add summary/embedding wiring and Opik integration, sentiment eval harness

### PR DESCRIPTION
**Pull request by rachellougee**
### What are the relevant tickets?

https://github.com/mitodl/ol-data-platform/issues/2544

### Description (What does it do?)

- **Opik integration**: LLM tracing, cost, and Prompt Library for the summarize/categorize/sentiment-classify calls (same Keycloak auth pattern as learn-ai)
- New dbt models: `int__feedback__summary`, `int__feedback__embedding`  — each gracefully handles its Dagster-written source table not existing yet.  
- Updates `afact_feedback_conversation` with real joins for summary and embedding data.
- **Sentiment eval harness** : one-time bake-off comparing embedding-kNN, a local classifier, and an LLM for sentiment scoring against the Zendesk-CSAT-labeled sample 

### How can this be tested?

Verified end-to-end against real data (Parley + real Zendesk conversations) in a personal dev schema, including Opik tracing/cost.

```
cd dg_projects/ml
export DAGSTER_ENV=dev
export DBT_SCHEMA_SUFFIX=<your_suffix>
export SUMMARY_PROVIDER=openai_compatible
export LLM_BASE_URL=https://parley.api.mit.edu/v1
export OPENAI_COMPATIBLE_API_KEY=<your key>

# Opik tracing/prompt library/cost (optional)
export OPIK_URL_OVERRIDE=https://opik-qa.ol.mit.edu/api/
export OPIK_PROJECT_NAME=dagster-ml
```

- Materialize `intermediate/feedback_summaries`, then `intermediate/feedback_embeddings`.
- `dbt build --target dev_production --vars 'schema_suffix: <your_suffix>' --select int__feedback__summary int__feedback__embedding afact_feedback_conversation`

- Sentiment eval harness
```yaml
ops:
  intermediate__feedback_sentiment_eval:
    config:
      include_llm: false
      embedding_model_version: <model you actually have embedded>
      embedding_dim: <matching dim>
```

- Opik (optional): after any LLM-calling asset above, check project `dagster-ml` for traces/cost and the Prompt Library at https://opik-qa.ol.mit.edu/default/projects/01a08cd3-7e1e-7649-a9ce-9af0a400e069

### CODE REVIEW

File: dg_projects/ml/ml/assets/feedback_approvals.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+RUN_STATUSES = {"promoted", "superseded"}
+
+
+class CategoryDecision(Config):
+    category_slug: str = Field(
+        description="The dim_feedback_category.category_slug being decided on."
+    )
+    category_status: str = Field(
+        description="The decision: approved, merged or deprecated."
+    )
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/assets/feedback_approvals.py]: Dropped 'merged' as an accepted category_status (feedback_approvals.py, _feedback_sources.yml, dim_feedback_categorys accepted_values test) rather than adding a merge-target field for now -- staged, pending commit.

File: src/ol_dbt/models/dimensional/afact_feedback_conversation.sql [hunk 2]
```diff
@@ -154,3 +165,9 @@ left join dominant_tag
     and turn_aggregates.feedback_source_fk = dominant_tag.feedback_source_fk
 left join feedback_category
     on dominant_tag.tag_slug = feedback_category.category_slug
+left join cluster_assignment
+    on conversation.feedback_conversation_pk = cluster_assignment.feedback_conversation_pk
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/afact_feedback_conversation.sql]: Added the promoted-cluster -> feedback_category_proposal -> dim_feedback_category(approved) join in afact_feedback_conversation.sql, COALESCEd ahead of the dominant_tag fallback. Verified against real data: an approved llm_discovered category now correctly assigns category_fk to its clusters conversations. Staged, pending commit.

File: src/ol_dbt/models/dimensional/dim_feedback_category.sql [hunk 3]
```diff
@@ -30,32 +35,62 @@ with ticket as (
...
     , combined.category_slug
     , min(combined.category_label) as category_label
     , cast(null as varchar) as category_parent_slug
-    , 'proposed' as category_status
-    , 'seed' as category_source
-    , cast(null as varchar) as cluster_run_id
+    -- coalesce, not a bare default: a slug with no approval row is still 'proposed'
+    , coalesce(min(approval.category_status), 'proposed') as category_status
+    , max(combined.category_source) as category_source
+    , max(combined.cluster_run_id) as cluster_run_id
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply src/ol_dbt/models/dimensional/dim_feedback_category.sql]: Fixed: dim_feedback_category now ranks combined rows per slug (seed precedence, then most-recently-updated) and pulls category_label/category_source/cluster_run_id from that one canonical row, instead of independent min()/max() picks that could mix provenance across rows. Staged, pending commit.

File: src/ol_dbt/models/intermediate/feedback/_feedback_sources.yml [hunk 4]
```diff
@@ -33,3 +33,126 @@ sources:
...
+    description: One row per clustering run (params, cluster/noise counts, silhouette),
+      materialized by the feedback_clustering Dagster asset. run_status here is whatever
+      the asset wrote (currently always 'proposed') -- promotion status comes from
+      feedback_cluster_run_promotion, not this column.
+    columns:
+    - name: cluster_run_id
+      description: str, uuid identifying this clustering run.
+      tests:
+      - not_null
+      - unique
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply src/ol_dbt/models/intermediate/feedback/_feedback_sources.yml]: Removed the column/table tests from the optional Dagster-written sources in _feedback_sources.yml (feedback_cluster_run, feedback_cluster_candidate, feedback_category_approval, feedback_cluster_run_promotion, feedback_category_proposal, feedback_summaries, feedback_embeddings) -- equivalent not_null/unique/accepted_values coverage already exists (or was added) on their guarded intermediate models instead. Staged, pending commit.

File: src/ol_dbt/models/intermediate/feedback/int__feedback__cluster_assignment.sql [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    , cast(null as integer) as cluster_id
+    , cast(null as double) as cluster_probability
+where false
+{% else %}
+with promotion as (
+    select
+        cluster_run_id
+        , row_number() over (order by promoted_at desc) as decision_rank
+    from {{ promotion_source }}
+    where run_status = 'promoted'
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply src/ol_dbt/models/intermediate/feedback/int__feedback__cluster_assignment.sql]: Fixed: int__feedback__cluster_assignment now ranks every decision (any status) per cluster_run_id by promoted_at before filtering, so a later superseded row correctly deactivates that run instead of being discarded before it can outrank the older promoted row. Staged, pending commit.

File: dg_projects/ml/ml/assets/feedback_approvals.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        msg = "decisions must have at least one entry."
+        raise Failure(msg)
+    invalid = sorted({d.category_status for d in config.decisions} - CATEGORY_STATUSES)
+    if invalid:
+        msg = (
+            f"category_status must be one of {sorted(CATEGORY_STATUSES)}, got "
+            f"invalid value(s): {invalid}"
+        )
+        raise Failure(msg)
+    approved_at = datetime.now(tz=UTC)
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/assets/feedback_approvals.py]: Added duplicate-category_slug rejection to feedback_category_approval (raises Failure listing the offending slugs) plus a regression test. Staged, pending commit.

File: dg_projects/ml/ml/assets/feedback_category_proposals.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+
+@asset(
+    code_version="feedback_category_proposals_v1",
+    group_name="feedback",
+    key=AssetKey(["intermediate", "feedback_category_proposal"]),
+    deps=[AssetKey(["dimensional", "afact_feedback_conversation"])],
+    io_manager_key="io_manager",
+    pool="feedback_category_proposals",
+    metadata={"schema": intermediate_database_name, "write_mode": "append"},
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/assets/feedback_category_proposals.py]: Switched feedback_category_proposals from append to upsert on (cluster_run_id, cluster_id), matching the documented one-row-per-cluster grain. Staged, pending commit.

File: dg_projects/ml/ml/assets/feedback_sentiment_eval.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    ).join(conversation_df, on="feedback_conversation_pk", how="left")
+    labeled_df = labeled_sentiment_sample(joined)
+
+    if labeled_df.height < 10:  # noqa: PLR2004 -- below this, a train/test split is meaningless
+        msg = (
+            f"Only {labeled_df.height} CSAT-labeled conversations have a "
+            f"matching embedding (model_version={embedding_model_version}, "
+            f"dim={embedding_dim}); too few for a meaningful eval."
+        )
+        raise Failure(msg)
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/assets/feedback_sentiment_eval.py]: Added a per-class-count check (Failure if fewer than 2 distinct sentiment labels) and made train_test_split_indices stratify per label instead of a plain shuffle, plus regression tests for both. Staged, pending commit.

File: dg_projects/ml/ml/lib/categorize.py [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    client = llm.get_client()
+    if isinstance(client, AnthropicBedrock):
+        return AnthropicCategoryLabelClient(
+            client, bedrock_model_version or BEDROCK_CATEGORY_MODEL_VERSION
+        )
+    if isinstance(client, Anthropic):
+        return AnthropicCategoryLabelClient(
+            client, model_version or CATEGORY_MODEL_VERSION
+        )
+    return OpenAICategoryLabelClient(client, model_version or CATEGORY_MODEL_VERSION)
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/lib/categorize.py]: Added the same client_class-aware Claude-model guard to OpenAICategoryLabelClient (mirroring OpenAISummaryClient), threaded through build_category_label_client via llm.client_class, plus tests for both the reject and openai_compatible-allow cases. Staged, pending commit.

File: dg_projects/ml/ml/lib/categorize.py [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+                "dominant_tags": ", ".join(inputs["dominant_tags"]),
+                "model_version": client.model_version,
+            }
+        )
+    if not rows:
+        return pl.DataFrame(
+            schema={
+                k: v for k, v in CATEGORY_PROPOSAL_SCHEMA.items() if k != "proposed_at"
+            }
+        )
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/lib/categorize.py]: Fixed: propose_categories' empty-result path was already schema-complete, but feedback_category_proposals.py only added proposed_at when height>0 -- now unconditional, so the cast to CATEGORY_PROPOSAL_SCHEMA succeeds on an empty result too. Staged, pending commit.

File: dg_projects/ml/ml/lib/sentiment_eval.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+    client = llm.get_client()
+    if isinstance(client, AnthropicBedrock):
+        return AnthropicSentimentClient(
+            client, bedrock_model_version or BEDROCK_SENTIMENT_MODEL_VERSION
+        )
+    if isinstance(client, Anthropic):
+        return AnthropicSentimentClient(
+            client, model_version or SENTIMENT_MODEL_VERSION
+        )
+    return OpenAISentimentClient(client, model_version or SENTIMENT_MODEL_VERSION)
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/lib/sentiment_eval.py]: Added the same client_class-aware Claude-model guard to OpenAISentimentClient (mirroring OpenAISummaryClient), threaded through build_sentiment_client, plus tests for both the reject and openai_compatible-allow cases. Staged, pending commit.

File: dg_projects/ml/ml/resources/opik_auth.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    configured for this environment.
+    """
+    template = default_template
+    client = get_opik_client()
+    if client is not None:
+        try:
+            prompt = client.get_prompt(name=name) or client.create_prompt(
+                name=name, prompt=default_template
+            )
+            template = prompt.prompt
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply dg_projects/ml/ml/resources/opik_auth.py]: Added get_prompt_version() to opik_auth.py and folded it into what feedback_summaries records as summary_model_version (model@promptversion) when Opik is configured -- so filter_unsummarized now re-submits a conversation after a Prompt Library edit, not just a model change. Scoped to feedback_summaries specifically, since it's the only place a stored version string gates re-processing (categorize.py/sentiment_eval.py have no equivalent skip-if-already-processed logic). Staged, pending commit.

**Comment by rachellougee** [comment]
Addressed all 12 review threads: fixed 11 (correctness bugs + missing guards/tests) and dropped 'merged' as a category_status (no merge-target field, deferred). Verified against real data (dbt build + a direct query confirming the cluster-to-approved-category join now assigns category_fk correctly) and the full ml_tests suite (129 passed)

**Review by blarghmatey — CHANGES_REQUESTED** [review]
## Architecture: drop run promotion

Assigning a conversation to a cluster shouldn't involve a person. With `feedback_cluster_run_promotion` as the only path from a clustering run to `afact_feedback_conversation`, every conversation embedded after the last promotion has no cluster until someone promotes a newer run. Each promotion also drops cluster-derived categories until proposals are re-run and re-approved. That guarantees a backlog and stale assignments as the corpus grows.

Cluster assignment should be continuous:

- place each newly embedded conversation into the live clusters on every embedding refresh
- re-cluster automatically on a schedule or trigger
- carry cluster identity across runs by membership overlap, so categories attach to a stable `cluster_key` instead of a run-local HDBSCAN integer

People still choose the clustering configuration (in code) and curate category labels. The spec revision is in #2671; the mechanics are in `feedback_ml_approach.md` §C.1.

Given that, I'd split this PR:

- **Land after the fixes below:** the summary/embedding wiring into `afact`, the Opik integration, and the sentiment eval.
- **Rework against the revised spec:** `feedback_cluster_run_promotion`, `int__feedback__cluster_assignment`, `feedback_category_proposals`, and the approval keying.

## Findings in the parts that would land

1. **A prompt-version flip re-summarizes the whole corpus** (`ml/resources/opik_auth.py:218`). `get_prompt_version` returns `"local"` both when the prompt doesn't exist in Opik yet and on any Opik error. On the first run with Opik enabled, rows get stamped `"local"`, and `render_prompt` then creates `v1`. The next run sees `"v1" != "local"` and re-summarizes every conversation, even though v1 is the same text. An Opik outage does it twice: once to `"local"`, then back. An unavailable version shouldn't count as a new version. Either skip the prompt-version check when the lookup fails, or record the version of the prompt that was actually rendered.
2. **The feedback source no longer strips `schema_suffix`** (`_feedback_sources.yml`). Every other source in the repo strips it. I haven't run a build, but with a suffix set, dev builds will read `feedback_redacted` from the developer's own schema, and `tfact_feedback.sql:23` reads that source with no existence guard.
3. **Native timestamps on `afact`** (`afact_feedback_conversation.sql:165`). `summarized_at` and `embedded_at` changed from ISO8601 varchar to timestamptz. `conversation_ingested_at` on the same table is still varchar, and the comment in `dim_feedback_category.sql` says this layer uses ISO8601 strings (it converts `proposed_at` for that reason).
4. **The `array(double)` cast only works on Trino** (`int__feedback__embedding.sql`). DuckDB rejects it with a parser error. `cross_db_functions.sql` already has a dispatched `null_double_array` macro for this case, and the cast needs the same treatment. CI takes the stub branch, so it won't catch this.
5. **The LLM client adapters are copied three times** (`ml/lib/summarize.py`, `ml/lib/categorize.py:134-252`, `ml/lib/sentiment_eval.py:164-274`). Only the prompt, `max_tokens` and response parsing differ between them. The Claude-id guard has already drifted between copies once, in the earlier review round.

6. **Deployed environments don't default to Bedrock** (`ml/definitions.py`). `llm` only picks Bedrock when `DAGSTER_ENV == "production"`, so QA and CI fall back to the Anthropic API. `embedding_llm` defaults to OpenAI everywhere. The Dagster deployment's env and env secrets carry no `ANTHROPIC_API_KEY` or `OPENAI_API_KEY`, while its IRSA role already grants `bedrock:InvokeModel*` in CI, QA and production. Default `llm` to `bedrock` and `embedding_llm` to `bedrock_embeddings` whenever `DAGSTER_ENV != "dev"`, and keep the env overrides for local dev. Every deployed environment then works without an API key, which also makes the Vault-fallback removal safe. `bedrock_embeddings` defaults to Titan Text Embeddings V2, which outputs the 1024 dimensions `EMBEDDING_DIM` expects. Which Bedrock embedding model to keep is still for the §B.1 bake-off to decide.

## Deployment

The Opik wiring for the `ml` code location (the `OPIK_*` env vars and Vault read access to `secret-operations/sso/opik`) is in mitodl/ol-infrastructure#5836. Tracing stays off in deployed Dagster until that merges.

## For the reworked part

These go away with promotion but are worth carrying into the redesign:

- Proposals derive "dominant existing tags" from `afact.category_fk`. After an approval, that's the approved cluster category, not the seed tag, so re-proposing feeds the previous label back to the LLM as context. Use the seed-tag assignment instead.
- Re-running proposals upserts over a cluster's row. A different LLM label produces a different slug, so a previously approved cluster silently loses its category.
- `feedback_category_approval` doesn't check that the slug exists in `dim_feedback_category`. A typo succeeds and does nothing.
- Approvals keyed by slug carry across runs. A new cluster whose label slugifies to an already-approved slug is assigned without anyone having looked at it. The revised spec keys categories on `cluster_key` instead.

**Comment by rachellougee** [comment]
Split done: cluster promotion, category proposals, and category approval all removed from this PR. Categories are seed-only again.

Findings 1–6 all fixed: prompt-version false-positive, schema_suffix fallback (shared `dev_schema_source` macro), ISO8601 timestamps, DuckDB array cast, LLM adapter dedup, Bedrock defaults for non-dev envs.

**Review by blarghmatey — APPROVED** [review]
